### ISSUE TITLE
Aborting a writable stream should not abort writes in progress

### ISSUE THREAD

**Issue by domenic** [whatwg/streams#611]
I am spinning this off from https://github.com/whatwg/streams/pull/607 so we can land that separately. The test in question is "Aborting a WritableStream prevents further writes after any that are in progress", which I have updated to expect that the write() fulfills, but currently it fails (if uncommented) with the aborted TypeError.

The problem is essentially that abort() synchronously rejects all pending write requests, whereas WritableStreamDefaultControllerProcessWrite does not unshift the pending write request until the underlying sink's write() call succeeds, which happens asynchronously.

We should instead unshift the pending write request the moment the underlying sink write() call is made, I think. This matches better the idea that once the underlying sink write() has started, it is entirely up to the underlying sink whether the writer.write() call should fulfill or reject. But a quick attempt to do that leads to lots of test timeouts and other failures. I would appreciate help investigating ways of making the test pass.

**Comment by ricea** [comment]
Meaning that we have an additional internal slot to hold the promise for the in-progress write()? It sounds good to me. I will have a go at getting it to work.

**Comment by ricea** [comment]
This is challenging.

I have a work-in-progress at https://github.com/ricea/streams/tree/in_progress_write but there are still some tests not passing.

Test 'Errors must be propagated backward: erroring via the controller errors a slow write' from error-propagation-backward.js times out, but I'm guessing that is expected from this change.

**Comment by domenic** [comment]
> Test 'Errors must be propagated backward: erroring via the controller errors a slow write' from error-propagation-backward.js times out, but I'm guessing that is expected from this change.

Oh, that is interesting. Probably it is OK, and we just say that if you want to signal a currently-ongoing write failed, you have to return a rejected promise; using controller.error() will only affect future writes.

**Comment by ricea** [comment]
My opinion is that "once a write() has dispatched, the returned status always depends on the underlying sink" is the easiest semantics to implement and understand.

The slightly surprising part is that write() may fulfill long after everything else has ended. This is the most informative behaviour, but also could be a trap for content authors expecting nothing else to happen after the .closed promise resolves. From a standard/implementation point of view, tracking "what kind of abort is this" adds considerable complexity, and may even be intractible. It's possible that whether a write() should succeed or fail in a particular scenario is subjective.

Additional peculiarity: normally write()s resolve in strict calling order, but on error one write() may resolve out-of-order.

If a sink is capable of cancelling an in-progress write on abort() then they should do so. I think platform-supplied sinks should always return a rejection if they successfully cancelled the underlying write operation or are unable to determine the outcome, and fulfill if the write operation completed. "reject on successful cancel" sounds strange but it's really "reject on write did not happen" which is the normal expected behaviour.

Underlying sinks provided by content authors can of course do whatever works for them, but hopefully will tend to match the platform behaviour.

I see two reasons for an underlying sink to implement abort():

1. To cancel the underlying write() operation, and/or
2. To signal the error condition to the underlying layer.

**Comment by domenic** [comment]
> My opinion is that "once a write() has dispatched, the returned status always depends on the underlying sink" is the easiest semantics to implement and understand.

That's not quite correct though, since the underlying sink could call `controller.error()`, but we plan to ignore that. It's more, "the returned status always depends on the return value of the underlying sink's `write()` method."

> This is the most informative behaviour, but also could be a trap for content authors expecting nothing else to happen after the .closed promise resolves.
> 
> Additional peculiarity: normally write()s resolve in strict calling order, but on error one write() may resolve out-of-order.

Ooh, these seem pretty bad :(. (I don't quite understand how the latter could arise, though.) Maybe we could delay .closed until the write() fulfills or rejects?

> "reject on successful cancel" sounds strange but it's really "reject on write did not happen" which is the normal expected behaviour.

Agreed, although I am not 100% sure about the unable-to-determine-the-outcome case.

**Comment by ricea** [comment]
> Ooh, these seem pretty bad :(. (I don't quite understand how the latter could arise, though.) Maybe we could delay .closed until the write() fulfills or rejects?

Consider

```javascript
writer.write('a').then(console.log('a')).catch(console.log('reject a'));
writer.write('b').then(console.log('b')).catch(console.log('reject b'));
writer.write('c').then(console.log('c')).catch(console.log('reject c'));
writer.ready.then(() => writer.abort());
```

Assume the underlying sink is slow.

With HWM=0 this logs
```
a
b
c
```

but with HWM=4 then it logs
```
reject b
reject c
a
```

I think in practice only the most careful people will put a .catch on write(), and so only the most careful people will see this happen. Since they are careful they will have proper tests and make sure it doesn't mess up their state machine.

Delaying .closed looks like a great idea. I will try implementing it and see how it comes out.

**Comment by domenic** [comment]
> but with HWM=4 then it logs

I'm not sure at what point things are just getting overcomplicated, but we could delay the whole erroring process---i.e. rejecting all pending writes, rejecting the closed promise, etc.---until the ongoing write finishes.

**Comment by ricea** [comment]
I have a working partial implementation at

https://github.com/ricea/streams/commit/d673b087f1478f9c7a1638e7ffc651d921a8f981

I haven't incorporated the new .closed semantics yet.

I expect we will refine the implementation, but it's worth looking at the changes to the tests to see if it makes sense.

**Comment by tyoshino** [comment]
I had been thinking that `.abort()` should make the WritableStream to mark everything on it including `.closed` , pending `.write()`s and active `.write()` as aborted. I.e. almost in the same expectation as `.cancel()`.

I don't object to the change. I agree that `.closed` should be delayed if we allow the active `.write()` to be fulfilled/rejected with the actual result of the corresponding `.write()` on the sink. Wait for the active `.write()`, reject all the pending `.write()`s and then `.closed`.

One more point we should clarify. If the in-flight operation is `.close()`, we `abort()`-ed but it finished successfully, I think we should move the WritableStream into `"closed"` state as well as `.write()`s fulfills regardless of `.abort()`. Do you agree?

**Comment by domenic** [comment]
> I had been thinking that .abort() should make the WritableStream to mark everything on it including .closed , pending .write()s and active .write() as aborted. I.e. almost in the same expectation as .cancel().

This is certainly simpler. But I think it is not as good for the producer, because they see .write() rejecting and everything shutting down immediately when in fact the underlying sink is still busy writing data to the filesystem or wherver. That is what motivates the change.

But I am still sympathetic to maybe not making this change if we just want to keep things simple. Maybe it will not impact people in the real world much. When they are aborting their writable streams maybe they do not care about whether or not the write succeeded.

> Wait for the active .write(), reject all the pending .write()s and then .closed.

Yeah, this sounds good.

> One more point we should clarify. If the in-flight operation is .close(), we abort()-ed but it finished successfully, I think we should move the WritableStream into "closed" state as well as .write()s fulfills regardless of .abort(). Do you agree?

Interesting. I think we discussed this before. We even have a note in the spec:

> Due to the way writable streams asynchronously close, it is possible for both close and abort to be called, in cases where the producer aborts the stream while it is in the "closing" state. Notably, since a stream always spends at least one turn in the "closing" state, code like ws.close(); ws.abort(...); will cause both to be called, even if the close method itself has no asynchronous behavior. A well-designed underlying sink object should be able to deal with this.

I think I agree with you that it'd be better to just disallow abort() during close(). But this note implies to me that we thought through this and came to the opposite conclusion in the past.  It looks like the discussions are in:

- https://github.com/whatwg/streams/issues/130
- https://github.com/whatwg/streams/pull/192
- https://github.com/whatwg/streams/pull/224
- See especially https://github.com/whatwg/streams/issues/130#issuecomment-53506871 proposals 1-3.

I guess the question is whether we think abort() should be able to abort a long-running underlying sink close (with no cooperation from the underlying sink). Probably we should have the same answer for a long-running underlying sink close as we do for a long-running underlying sink write.

---

I guess at this point I am on the fence as to whether this change is worth making. I will review @ricea's branch on the assumption that it is. But if anyone has strong feelings about how abort() should behave, that would be helpful, as I am no longer sure myself.

**Comment by tyoshino** [comment]
> But I think it is not as good for the producer, because they see .write() rejecting and everything shutting down immediately when in fact the underlying sink is still busy writing data to the filesystem or wherver. That is what motivates the change.

I'm ok with that. After several discussions, we have consensus that write() promise should represent the ongoing operation. Yes.

> I think I agree with you that it'd be better to just disallow abort() during close().

Oh, no. Sorry if the first sentence of my post gave some nuance that I'm stating objection also in the third sentence. I just wanted to say that it seems we should align `.closed`'s transition for `abort()` on ongoing `close()` with the proposed behavior of `abort()` with ongoing `write()`. I.e. if `underlyingSink.close()` says it was successful even though it also received `underlyingSink.abort()` call, we should fulfill `.closed` with undefined even if `.abort()` was called while `.close()` is ongoing. I didn't mean that we should ignore `abort()` when there's an ongoing `close()`.

**Comment by tyoshino** [comment]
Hmm, it seems you're revisiting the point (allow abort()-ing ongoing close()) based on findings beyond what I was thinking. Could you please elaborate your concerns?

**Comment by ricea** [comment]
The example I have been thinking about is an application which reads a resource from the network, checks it for validity, and appends it to a file. If the resource is invalid, it aborts, truncates the file to where it was to start with, and retries. It looks like this:

```javascript
fetch(resource)
  .then(response -> response.body)
  .pipeThrough(new TransformStream(validator))
  .pipeTo(fileStream)
  .catch(truncateFileAndRetry);
```

The problem here is that if the pipeTo() promise rejects before writes by the underlying sink of `fileStream` have finished, then those writes will race with the file truncation.

Without these changes, the underlying sink can ensure deterministic behaviour by preventing abort() from fulfilling until any underlying operations are complete. However, this requirement is non-obvious and requires extra book-keeping inside the sink implementation.

With the changes to write(), the implementer of the underlying sink would have to go out of their way to get this wrong.

The reverse issue is when the user of the WritableStream is just discarding the sink and just wants abort() to be instant. Since they can easily implement this themselves by calling abort() and setting some flag to disregard any promises that settle after that point, I don't see this as a significant problem.

Unfortunately, this doesn't throw light on the close() vs. abort() issue because pipeTo() won't abort() once close() has started.

**Comment by ricea** [comment]
https://github.com/whatwg/streams/commit/73d979e7d488b643cc42576b6ec4af2cd7cf29cc changes the timing of .closed so it doesn't happen before write() when you abort(). I do not think it is unreasonably complex.

I plan to do the same thing for queued calls to write() next. It should be similarly straightforward.

**Comment by ricea** [comment]
Queued calls to write() done. I have created a pull request to iterate on the implementation: #619.

**Comment by tyoshino** [comment]
> If the resource is invalid, it aborts, truncates the file to where it was to start with, and retries. It looks like this:

Great example for analysis. But the promise returned by the `pipeTo()` call rejects after `underlyingSink.abort()` fulfills. So, this issue could be addressed by delaying the fulfillment.

That said, we can say the same for `.closed`. We cannot determine the timing when we can perform an operation like `truncateFileAndRetry` by `.closed` without the proposed change.

> Since they can easily implement this themselves by calling abort() and setting some flag to disregard any promises that settle after that point, I don't see this as a significant problem.

Agreed. I also quickly looked for any use case where `.closed` is expected to be rejected immediately, but I don't see anything.

**Comment by domenic** [comment]
This was fixed in #619.
